core-addon/IonCore.js

```diff
@@ -14,11 +17,153 @@ module.exports = class IonCore {
       }
       this._openControlPanel();
     });
+
+    // Listen for messages from the options page.
+    browser.runtime.onMessage.addListener(
+      (m, s) => this._handleMessage(m, s));
   }
 
   _openControlPanel() {
     browser.runtime.openOptionsPage().catch(e => {
       console.error(`IonCore.js - Unable to open the control panel`, e);
     });
   }
+
+  /**
+   * Handles messages coming in from the options page.
+   *
+   * @param {Object} message
+   *        The payload of the message.
+   * @param {runtime.MessageSender} sender
+   *        An object containing informations about who sent
+   *        the message.
+   * @returns {Promise} The response to the received message.
+   *          It can be resolved with a value that is sent to the
+   *          `sender`.
+   */
+  _handleMessage(message, sender) {
+    // We only expect messages coming from the embedded options page
+    // at this time. Discard anything else and report an error.
+    if (sender.url != browser.runtime.getURL(ION_OPTIONS_PAGE_PATH)) {
+      return Promise.reject(
+        new Error("IonCore - received message from unexpected sender"));
+    }
+
+    switch (message.type) {
+      case "enrollment": {
+        // Let's not forget to respond `true` to the sender: the UI
+        // is expecting it.
+        return this._enroll().then(r => true);
+      } break;
+      case "study-enrollment": {
+        // Let's not forget to respond `true` to the sender: the UI
+        // is expecting it.
+        return this._enrollStudy(message.data.studyID).then(r => true);
+      } break;
+      default:
+        return Promise.reject(
+          new Error(`IonCore - unexpected message type ${message.type}`));
+    }
+  }
+
+  /**
+   * Enroll in the Ion platform.
+   *
+   * This sets up all the required information (e.g. Ion ID)
+   * and sets the relevant data to the pipeline.
+   *
+   * @returns {Promise} A promise resolved when the enrollment
+   *          is complete (does not block on data upload).
+   */
+  async _enroll() {
+    // Generate a proper random UUID.
+    const uuid = await browser.legacyTelemetryApi.generateUUID();
+
+    // Store it locally for future use.
+    await browser.storage.local.set({ionId: uuid});
+
+    // The telemetry API, before sending a ping, reads the
+    // ion id from a pref. It no value is set, the API will
+    // throw and nothing will be sent. This means, at enrollment,
+    // we need set the value of that required pref.
+    await browser.legacyTelemetryApi.setIonID(uuid);
+
+    // Finally send the ping.
+    await this._sendEnrollmentPing();
+  }
+
+  /**
+   * Enroll in an Ion Study.
+   *
+   * This sends the required pings,
+   *
+   * @returns {Promise} A promise resolved when the enrollment
+   *          is complete (does not block on data upload).
+   */
+  async _enrollStudy(studyAddonId) {
+    // TODO: Validate the study id?
```

Comment:
This is a good idea... maybe just that it's a valid UUID

Reply:
Yes! Are addon ids always uuids? I think I saw things such as "addon@mozilla.com"